### PR TITLE
Receive rules for LoadedRules directly from LogicalPlanner

### PR DESCRIPTION
**Pull request by mkleen**
This removes the rules from `LoadedRules` and uses the rules directly from the optimizers from the `LogicalPlanner` to avoid inconsistencies.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java [hunk 1]
```diff
@@ -42,7 +42,7 @@ public class CoordinatorSessionSettings extends SessionSettings {
 
     private final User authenticatedUser;
     private User sessionUser;
-    private Set<Class<? extends Rule<?>>> excludedOptimizerRules;
+    private Set<Class<? extends Rule>> excludedOptimizerRules;
```
Comment (mkleen) [inline server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java:45]: `rule.getClass()`only returns `Class<? extends Rule>` instead of `Class<? extends Rule<?>>` without a cast. This makes no difference.

**Comment by mkleen** [comment]
Ready for review again, now with the second approach.

File: server/src/main/java/io/crate/planner/optimizer/LoadedRules.java [hunk 2]
```diff
@@ -32,71 +32,21 @@
...
-
     @Override
     public List<SessionSetting<?>> sessionSettings() {
-        return Lists2.map(rules, this::buildRuleSessionSetting);
+        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting(x.getClass()));
     }
 
     @VisibleForTesting
-    SessionSetting<?> buildRuleSessionSetting(Class<? extends Rule<?>> rule) {
+    SessionSetting<?> buildRuleSessionSetting(Class<? extends Rule> rule) {
```
Comment (mkleen) [inline server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:49]: `rule.getClass` yields `Class<? extends Rule>` instead of `Class<? extends Rule<?>>` from `Rule.class`. Makes no difference in Practice.

File: server/src/main/java/io/crate/planner/operators/LogicalPlanner.java [hunk 3]
```diff
@@ -600,4 +572,49 @@ protected LogicalPlan visitAnalyzedInsertStatement(AnalyzedInsertStatement state
...
+    private static final List<Rule<?>> WRITE_OPTIMIZER_RULES =
+        List.of(new RewriteInsertFromSubQueryToInsertFromValues());
+
+    public static List<Rule<?>> optimizerRules() {
+        var result = new ArrayList<Rule<?>>();
+        result.addAll(ITERATIVE_OPTIMIZER_RULES);
+        result.addAll(FETCH_OPTIMIZER_RULES);
+        result.addAll(WRITE_OPTIMIZER_RULES);
+        return result;
+    }
```
Comment (seut) [inline server/src/main/java/io/crate/planner/operators/LogicalPlanner.java:619]: Looks like that this is only used inside `LoadedRules` so better move there?

File: server/src/main/java/io/crate/planner/operators/LogicalPlanner.java [hunk 4]
```diff
@@ -600,4 +572,49 @@ protected LogicalPlan visitAnalyzedInsertStatement(AnalyzedInsertStatement state
...
+        new RewriteNestedLoopJoinToHashJoin()
+    );
+
+    private static final List<Rule<?>> FETCH_OPTIMIZER_RULES = List.of(
+        new RemoveRedundantFetchOrEval(),
+        new RewriteToQueryThenFetch()
+    );
+
+    private static final List<Rule<?>> WRITE_OPTIMIZER_RULES =
+        List.of(new RewriteInsertFromSubQueryToInsertFromValues());
```
Comment (seut) [inline server/src/main/java/io/crate/planner/operators/LogicalPlanner.java:611]: I'd prefer declaring static definitions on top of the class, especially as these are used inside the ctor.

File: server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java (hunk 1 above)
Comment (seut) [inline server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java:45]: Why did you change these declarations? We shouldn't use raw parameterized classes. Actually we should drive the opposite: fix any raw parameterized usages when spotted.

**Review by seut — COMMENTED** [review]
Good catch! I've added some suggestions.

File: server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java (hunk 1 above)
Comment (mkleen) [reply server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java:45]: See the comment for this.

Comment (mkleen) [reply server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java:45]: The only option would be a cast to `Class<? extends Rule<?>`.

File: server/src/main/java/io/crate/planner/optimizer/LoadedRules.java (hunk 2 above)
Comment (seut) [reply server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:49]: Could be solved by:
```suggestion
    SessionSetting<?> buildRuleSessionSetting(Rule<?> rule) {
        var clazz = rule.getClass();
        var simpleName = clazz.getSimpleName();
        ....
```

File: server/src/main/java/io/crate/planner/optimizer/LoadedRules.java [hunk 5]
```diff
@@ -32,71 +32,21 @@
...
-        RewriteGroupByKeysLimitToLimitDistinct.class,
-        RewriteInsertFromSubQueryToInsertFromValues.class,
-        RewriteToQueryThenFetch.class,
-        RewriteNestedLoopJoinToHashJoin.class
-    );
-
     @Override
     public List<SessionSetting<?>> sessionSettings() {
-        return Lists2.map(rules, this::buildRuleSessionSetting);
+        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting(x.getClass()));
```
Comment (seut) [inline server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:45]: When declaring a static list inside this class as already suggested:
```
    private static final List<Rule<?>> RULES = Lists2.concat(
            Lists2.concat(
                    ITERATIVE_OPTIMIZER_RULES,
                    FETCH_OPTIMIZER_RULES
            ),
            WRITE_OPTIMIZER_RULES
    );
```

and the suggestion for the `buildRuleSessionSetting`, this could stay similar:
```
        return Lists2.map(RULES, this::buildRuleSessionSetting);
```

File: server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java (hunk 1 above)
Comment (seut) [reply server/src/main/java/io/crate/metadata/settings/CoordinatorSessionSettings.java:45]: Ok I understood the reason by reading your comment (Sorry slipped through). But as stated, we shouldn't change it into raw parameterized declaration, but rather change the accessing code like suggested in my other comments.

File: server/src/main/java/io/crate/planner/optimizer/LoadedRules.java (hunk 2 above)
Comment (mkleen) [reply server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:49]: As mentioned above, `rule.getClass()` will return `Class<? extends Rule>` not  `Class<? extends Rule<?>>`, therefore this will not resolve it.

Comment (mfussenegger) [reply server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:49]: I think it would still make sense to at least limit the warning to the place where it's an issue.

E.g. do something like:

```diff

     @Override
+    @SuppressWarnings("unchecked")
     public List<SessionSetting<?>> sessionSettings() {
-        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting(x.getClass()));
+        return Lists2.map(LogicalPlanner.optimizerRules(), x -> buildRuleSessionSetting((Class<? extends Rule<?>>) x.getClass()));
     }
```

And keep using `Class<? extends Rule<?>> rule` everywhere else. Otherwise the warnings are scattered all over the place.


> What is the difference between Rule<?> and Rule ? Rule<?> expands to Rule<? extends Object> and Rule expands to Rule<Object>. The difference in the boundary of these types are minimal.

In this case it may not matter much, but there is a difference between `Rule<?>` and a raw `Rule`. In the latter the type erasue also maps to methods. An example:

```
    class Foo<T> {

        private List<String> names() {
            return List.of("x");
        }
    }

    @Test
    public void test_foo() throws Exception {
        Foo x = new Foo();

        // x.names() is _not_ List<String> because Foo is a raw type
        List names = x.names();
    }
```

Comment (mkleen) [reply server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:49]: Thanks for the feedback, let's go with a cast then.

Comment (mkleen) [reply server/src/main/java/io/crate/planner/optimizer/LoadedRules.java:49]: Btw, I was indeed not aware that the raw type would also erase the types of the methods generics.

**Comment by mkleen** [comment]
retest this please

**Comment by mkleen** [comment]
Could maybe one of the other reviewers take over, please ? I am depending on this in another pr and would like to get it merged.

**Review by mfussenegger — APPROVED** [review]
